bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/text/AbstractRuleBasedInterpreter.java

```diff
@@ -538,52 +589,73 @@ protected String executeSingle(ResourceBundle language, String[] labelFragments,
      * @param language Language information that is used for matching
      * @param labelFragments label fragments that are used to match an item's label.
      *            For a positive match, the item's label has to contain every fragment - independently of their order.
-     *            They are treated case insensitive.
+     *            They are treated case-insensitive.
      * @param commandType optional command type that all items have to support.
      *            Provide {null} if there is no need for a certain command to be supported.
      * @return All matching items from the item registry.
      */
     protected List<Item> getMatchingItems(ResourceBundle language, String[] labelFragments,
             @Nullable Class<?> commandType) {
-        List<Item> items = new ArrayList<>();
-        Map<Item, List<Set<String>>> map = getItemTokens(language.getLocale());
-        for (Item item : map.keySet()) {
-            for (Set<String> parts : map.get(item)) {
-                boolean allMatch = true;
-                for (String fragment : labelFragments) {
-                    if (!parts.contains(fragment.toLowerCase(language.getLocale()))) {
-                        allMatch = false;
+        Set<Item> items = new HashSet<>();
+        Set<Item> exactMatchItems = new HashSet<>();
+        Map<Item, List<List<List<String>>>> map = getItemTokens(language.getLocale());
+        for (Entry<Item, List<List<List<String>>>> entry : map.entrySet()) {
+            Item item = entry.getKey();
+            for (List<List<String>> parts : entry.getValue()) {
+                boolean exactMatch = false;
+                logger.trace("Checking tokens {} against the item tokens {}", labelFragments, parts);
+                List<String> lowercaseLabelFragments = Arrays.stream(labelFragments)
+                        .map(lf -> lf.toLowerCase(language.getLocale())).toList();
+                List<String> unmatchedFragments = new ArrayList<>(lowercaseLabelFragments);
+                for (List<String> segment : parts) {
+                    if (segment.equals(lowercaseLabelFragments)) {
+                        exactMatch = true;
```

Comment:
Why is that important? And I believe this is not true in general:

- labelFragments `[a, b, c, d]`
- segments: `[[b], [c, d], [a]]`
- first segment, there is `[a, c, d]` left
- second segment, there is `[a]` left
- third segment matches everything that is left, so `exactMatch = true`

Is that intended?

Reply:
Thank you for the review, and sorry for taking so long to answer, it caught me in the middle of another task that took more than expected.

Maybe I don't understanding the question but I think the code ok.
Each segment is compared against lowercaseLabelFragments which is never modified, checking if some segment in the path is an exact match agains the label extracted from the voice command. Each iteration I'm trying to remove from unmatchedFragments but that is a different list initialized with same strings.

I have renamed the segment and parts variables trying to make that part easier to read.